File: common/src/main/java/com/ssblur/scriptor/item/casters/CoordinateCasterCrystal.java
package com.ssblur.scriptor.item.casters;

import com.ssblur.scriptor.data_components.ScriptorDataComponents;
import com.ssblur.scriptor.events.network.server.ServerTraceNetwork;
import com.ssblur.scriptor.helpers.ComponentHelper;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import net.minecraft.ChatFormatting;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.Level;
import org.apache.commons.lang3.tuple.Pair;

import java.util.ArrayList;
import java.util.List;

public class CoordinateCasterCrystal extends CasterCrystal {
  public CoordinateCasterCrystal(Properties properties) {
    super(properties);
  }

  public static class BlockPosDirection extends Pair<BlockPos, Direction> {
    BlockPos left;
    Direction right;
    public BlockPosDirection(BlockPos left, Direction right) {
      this.left = left;
      this.right = right;
    }

    @Override
    public BlockPos getLeft() {
      return left;
    }

    @Override
    public Direction getRight() {
      return right;
    }

    @Override
    public Direction setValue(Direction value) {
      return right = value;
    }
  }

  @Override
  public List<Targetable> getTargetables(ItemStack itemStack, Level level) {
    ArrayList<Targetable> list = new ArrayList<>();
    var coordinates = itemStack.get(ScriptorDataComponents.COORDINATES);
    if(coordinates != null)
      for(var pos: coordinates) {
        var targetable = new Targetable(level, new BlockPos(pos.get(0), pos.get(1), pos.get(2)));
        targetable.setFacing(Direction.values()[pos.get(3)]);
        list.add(targetable);
      }
    return list;
  }

  @Override
  public void appendHoverText(ItemStack itemStack, TooltipContext level, List<Component> list, TooltipFlag tooltipFlag) {
    super.appendHoverText(itemStack, level, list, tooltipFlag);
    var coordinates = getCoordinates(itemStack);
    for(var pair: coordinates) {
      var coordinate = pair.getLeft();
      ComponentHelper.updateTooltipWith(ChatFormatting.GRAY, list, "lore.scriptor.coordinate_crystal_3", coordinate.getX(), coordinate.getY(), coordinate.getZ());
    }

    if(coordinates.isEmpty())
      list.add(Component.translatable("lore.scriptor.coordinate_crystal_1").withStyle(ChatFormatting.GRAY));
    else {
      if(coordinates.size() < 4)
        list.add(Component.translatable("lore.scriptor.coordinate_crystal_2").withStyle(ChatFormatting.GRAY));
      list.add(Component.translatable("lore.scriptor.crystal_reset").withStyle(ChatFormatting.GRAY));
    }
  }

  @Override
  public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand interactionHand) {
    var result = super.use(level, player, interactionHand);

    if(!level.isClientSide) {
      ItemStack itemStack = player.getItemInHand(interactionHand);
      ServerTraceNetwork.requestTraceData(player, target -> addCoordinate(itemStack, target.getTargetBlockPos(), target.getFacing()));
    }

    return result;
  }

  public static void addCoordinate(ItemStack itemStack, BlockPos pos, Direction direction) {
    var list = itemStack.get(ScriptorDataComponents.COORDINATES);
    if(list == null)
      list = new ArrayList<>();
    if(list.size() < 4) {
      list = new ArrayList<>(list);
      list.add(List.of(pos.getX(), pos.getY(), pos.getZ(), direction.ordinal()));
    }
    itemStack.set(ScriptorDataComponents.COORDINATES, list);
  }

  public static List<BlockPosDirection> getCoordinates(ItemStack itemStack) {
    ArrayList<BlockPosDirection> list = new ArrayList<>();
    var coordinates = itemStack.get(ScriptorDataComponents.COORDINATES);
    if(coordinates != null)
      for(var pos: coordinates)
        list.add(new BlockPosDirection(new BlockPos(pos.get(0), pos.get(1), pos.get(2)), Direction.values()[pos.get(3)]));
    return list;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/item/casters/CoordinateCasterCrystal.kt
package com.ssblur.scriptor.item.casters

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.PoseStack
import com.ssblur.scriptor.block.ScriptorBlocks
import com.ssblur.scriptor.data.components.ScriptorDataComponents
import com.ssblur.scriptor.helpers.ComponentHelper
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.item.interfaces.ItemWithCustomRenderer
import com.ssblur.scriptor.network.server.TraceNetwork
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.client.player.AbstractClientPlayer
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.network.chat.Component
import net.minecraft.util.Mth
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResultHolder
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.DirectionalBlock
import org.apache.commons.lang3.tuple.Pair
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

class CoordinateCasterCrystal(properties: Properties): CasterCrystal(properties), ItemWithCustomRenderer {
  class BlockPosDirection(var blockPos: BlockPos, var direction: Direction): Pair<BlockPos, Direction>() {
    override fun getLeft(): BlockPos {
      return blockPos
    }

    override fun getRight(): Direction {
      return direction
    }

    override fun setValue(newValue: Direction): Direction {
      return newValue.also { direction = it }
    }

    fun setValue(newValue: BlockPos): BlockPos {
      return newValue.also { blockPos = it }
    }
  }

  override fun getTargetables(itemStack: ItemStack?, level: Level?): List<Targetable?> {
    val list = ArrayList<Targetable?>()
    val coordinates = itemStack!!.get(ScriptorDataComponents.COORDINATES)
    if (coordinates != null) for (pos in coordinates) {
      val targetable = Targetable(
        level!!,
        BlockPos(pos[0], pos[1], pos[2])
      )
      targetable.setFacing(Direction.entries[pos[3]])
      list.add(targetable)
    }
    return list
  }

  override fun appendHoverText(
    itemStack: ItemStack,
    level: TooltipContext,
    list: MutableList<Component>,
    tooltipFlag: TooltipFlag
  ) {
    super.appendHoverText(itemStack, level, list, tooltipFlag)
    val coordinates = getCoordinates(itemStack)
    for (pair in coordinates) {
      val coordinate = pair.getLeft()
      ComponentHelper.updateTooltipWith(
        ChatFormatting.GRAY,
        list,
        "lore.scriptor.coordinate_crystal_3",
        coordinate.x,
        coordinate.y,
        coordinate.z
      )
    }

    if (coordinates.isEmpty()) list.add(
      Component.translatable("lore.scriptor.coordinate_crystal_1").withStyle(
        ChatFormatting.GRAY
      )
    )
    else {
      if (coordinates.size < 4) list.add(
        Component.translatable("lore.scriptor.coordinate_crystal_2").withStyle(
          ChatFormatting.GRAY
        )
      )
      list.add(Component.translatable("lore.scriptor.crystal_reset").withStyle(ChatFormatting.GRAY))
    }

    list.add(Component.translatable("lore.scriptor.crystal_focus").withStyle(ChatFormatting.GRAY))

  }

  override fun use(
    level: Level,
    player: Player,
    interactionHand: InteractionHand
  ): InteractionResultHolder<ItemStack> {
    val result = super.use(level, player, interactionHand)

    if (!level.isClientSide) {
      val itemStack = player.getItemInHand(interactionHand)
      TraceNetwork.requestTraceData(player, false) { target: Targetable ->
        addCoordinate(
          itemStack,
          target.targetBlockPos,
          target.facing
        )
      }
    }

    return result
  }

  companion object {
    fun addCoordinate(itemStack: ItemStack, pos: BlockPos, direction: Direction) {
      var list = itemStack.get(ScriptorDataComponents.COORDINATES)
      if (list == null) list = ArrayList()
      if (list.size < 4) {
        list = ArrayList(list)
        list.add(java.util.List.of(pos.x, pos.y, pos.z, direction.ordinal))
      }
      itemStack.set(ScriptorDataComponents.COORDINATES, list)
    }

    @JvmStatic
    fun getCoordinates(itemStack: ItemStack): List<BlockPosDirection> {
      val list = ArrayList<BlockPosDirection>()
      val coordinates = itemStack.get(ScriptorDataComponents.COORDINATES)
      if (coordinates != null) for (pos in coordinates) list.add(
        BlockPosDirection(
          BlockPos(
            pos[0],
            pos[1],
            pos[2]
          ), Direction.entries[pos[3]]
        )
      )
      return list
    }
  }

  override fun render(
    player: AbstractClientPlayer,
    i: Float,
    pitch: Float,
    hand: InteractionHand,
    swingProgress: Float,
    itemStack: ItemStack,
    readyProgress: Float,
    matrix: PoseStack?,
    buffer: MultiBufferSource,
    lightLevel: Int
  ): Boolean {
    if(matrix == null) return false
    val minecraft = Minecraft.getInstance()
    val dispatcher = minecraft.blockRenderer
    val gameRenderer = minecraft.gameRenderer
    val camera = gameRenderer.mainCamera.position
    matrix.pushPose()
    matrix.setIdentity()
    val g = player.walkDist - player.walkDistO
    val h = -(player.walkDist + g * i).toDouble()
    val j = Mth.lerp(i, player.oBob, player.bob)
    matrix.translate(
      (sin(h * Math.PI) * j * 0.5),
      -abs((cos(h * Math.PI) * j)),
      0.0
    )
    matrix.translate(-camera.x, -camera.y, -camera.z)

    RenderSystem.disableDepthTest()
    RenderSystem.enableBlend()
    RenderSystem.setShader { GameRenderer.getPositionTexColorShader() }
    RenderSystem.disableCull()

    for ((pos, dir) in getCoordinates(itemStack)) {
      matrix.pushPose()
      val blockState = ScriptorBlocks.HIGHLIGHT_MODEL.get().defaultBlockState().setValue(DirectionalBlock.FACING, dir)
      matrix.translate(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())
      dispatcher.renderSingleBlock(blockState, matrix, buffer, 0xffffff, 0)
      matrix.popPose()
    }

    matrix.popPose()
    return false
  }
}
